agent/collector/http_server.go

```diff
@@ -0,0 +1,204 @@
+package collector
+
+import (
+	"bytes"
+	"compress/gzip"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"strings"
+
+	"github.com/gorilla/handlers"
+	"github.com/gorilla/mux"
+
+	"go.opentelemetry.io/collector/pdata/ptrace/ptraceotlp"
+	pb "go.opentelemetry.io/proto/otlp/collector/trace/v1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	protoBufContentType = "application/x-protobuf"
+	jsonContentType     = "application/json"
+)
+
+type httpServer struct {
+	addr     string
+	ingester ingester
+
+	hServer *http.Server
+}
+
+func newHttpServer(addr string, ingester ingester) *httpServer {
+	return &httpServer{
+		addr:     addr,
+		ingester: ingester,
+	}
+}
+
+func (s *httpServer) Start() error {
+	r := mux.NewRouter()
+	r.HandleFunc("/v1/traces", s.Export).Methods("POST")
+
+	s.hServer = &http.Server{
+		Addr:    s.addr,
+		Handler: handlers.CompressHandler(decompressBodyHandler(handlers.ContentTypeHandler(r, protoBufContentType, jsonContentType))),
+	}
+	listener, err := net.Listen("tcp", s.addr)
+	if err != nil {
+		return fmt.Errorf("cannot listen on address %s: %w", s.addr, err)
+	}
+
+	return s.hServer.Serve(listener)
+}
+
+func (s *httpServer) Stop() {
+	s.hServer.Close()
+}
+
+func (s httpServer) Export(w http.ResponseWriter, r *http.Request) {
+	contentType := r.Header.Get("content-type")
+	response := newHttpResponse(w, contentType)
+
+	request, err := s.parseBody(r.Body, contentType)
+	if err != nil {
+		response.sendError(http.StatusUnprocessableEntity, status.Errorf(codes.InvalidArgument, "Could not parse request body %s", err.Error()))
+		return
+	}
+
+	result, err := s.ingester.Ingest(r.Context(), request, "HTTP")
+	if err != nil {
+		response.sendError(http.StatusInternalServerError, status.Errorf(codes.InvalidArgument, "Error when ingesting spans %s", err.Error()))
+		return
+	}
+
+	response.send(http.StatusOK, result)
+}
+
+func (s httpServer) parseProtoBuf(body []byte) (*pb.ExportTraceServiceRequest, error) {
+	request := pb.ExportTraceServiceRequest{}
+
+	err := proto.Unmarshal(body, &request)
+	if err != nil {
+		return nil, err
+	}
+
+	return &request, nil
+}
+
+func (s httpServer) parseJson(body []byte) (*pb.ExportTraceServiceRequest, error) {
+	exportRequest := ptraceotlp.NewRequest()
+
+	err := exportRequest.UnmarshalJSON(body)
+	if err != nil {
+		return nil, err
+	}
+
+	protoBody, err := exportRequest.MarshalProto()
+	if err != nil {
+		return nil, err
+	}
+
+	return s.parseProtoBuf(protoBody)
+}
+
+func (s httpServer) parseBody(reqBody io.ReadCloser, contentType string) (*pb.ExportTraceServiceRequest, error) {
+	var body []byte
+	if b, err := io.ReadAll(reqBody); err == nil {
+		body = b
+	} else {
+		return nil, err
+	}
```

Comment:
Is there a reason to use this `if/else`, or can we simplify it using an early return, like this?
```suggestion
	var body []byte
	body, err := io.ReadAll(reqBody)
	
	if err != nil {
		return nil, err
	}
```